drivers/sensor/bosch/bmp180/bmp180.c

```diff
@@ -0,0 +1,463 @@
+					enum sensor_channel chan, uint16_t val)
+{
+	struct bmp180_data *data = dev->data;
+
+	/* Value must be a positive value 0-3 */
+	if ((chan != SENSOR_CHAN_PRESS) || (val > 3)) {
+		return -EINVAL;
+	}
+
+	data->osr_pressure = val;
+
+	return 0;
+}
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+static int bmp180_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif /* CONFIG_PM_DEVICE */
+
+	switch (attr) {
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+	case SENSOR_ATTR_OVERSAMPLING:
+		ret = bmp180_attr_set_oversampling(dev, chan, val->val1);
+		break;
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+
+static int read_raw_temperature(const struct device *dev)
+{
+	int ret;
+	uint8_t ctrlreg;
+	uint8_t reg16[2];
+	struct bmp180_data *data = dev->data;
+
+	/* send temperature measurement command */
+	ret = bmp180_reg_write(dev, BMP180_REG_MEAS_CTRL, BMP180_CMD_GET_TEMPERATURE);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * for the first while read get_temp_delay+1ms which is the convension time
+	 * descripted in the data-sheet in case the register not yet ready wait again
+	 */
+	k_sleep(K_MSEC(BMP180_CMD_GET_TEMP_DELAY));
+
+	while (true) {
+		k_sleep(K_MSEC(1));
+		ret = bmp180_reg_read(dev, BMP180_REG_MEAS_CTRL, &ctrlreg, 1);
+		if (ret) {
+			return ret;
+		}
+
+		/* bit 5 is 1 until data registers are ready */
+		if ((ctrlreg & BMP180_STATUS_CMD_RDY) == 0) {
+			break;
+		}
+	}
+
+	/* read msb and lsb */
+	ret = bmp180_reg_read(dev, BMP180_REG_MSB, (uint8_t *)&reg16, 2);
+	if (ret) {
+		return ret;
+	}
+
+	data->raw_temp = (int32_t)(sys_get_be16(reg16));
+	data->comp_temp = 0;
+
+	return 0;
+}
+
+static int read_raw_pressure(const struct device *dev)
+{
+	int ret;
+	uint8_t ctrlreg;
+	uint8_t reg24[3];
+	uint32_t delay;
+	int32_t pressure;
+	struct bmp180_data *data = dev->data;
+
+	switch (data->osr_pressure) {
+	case BMP180_ULTRALOWPOWER:
+		ctrlreg = BMP180_CMD_GET_OSS0_PRESS;
+		delay = BMP180_CMD_GET_OSS0_DELAY;
+		break;
+	case BMP180_STANDARD:
+		ctrlreg = BMP180_CMD_GET_OSS1_PRESS;
+		delay = BMP180_CMD_GET_OSS1_DELAY;
+		break;
+	case BMP180_HIGHRES:
+		ctrlreg = BMP180_CMD_GET_OSS2_PRESS;
+		delay = BMP180_CMD_GET_OSS2_DELAY;
+		break;
+	case BMP180_ULTRAHIGH:
+		ctrlreg = BMP180_CMD_GET_OSS3_PRESS;
+		delay = BMP180_CMD_GET_OSS3_DELAY;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = bmp180_reg_write(dev, BMP180_REG_MEAS_CTRL, ctrlreg);
+	if (ret) {
+		return ret;
+	}
+
+	k_sleep(K_MSEC(delay));
+
+	while (true) {
+		k_sleep(K_MSEC(1));
+		ret = bmp180_reg_read(dev, BMP180_REG_MEAS_CTRL, &ctrlreg, 1);
+		if (ret) {
+			return ret;
+		}
+
+		if ((ctrlreg & BMP180_STATUS_CMD_RDY) == 0) {
+			break;
+		}
+	}
```

Comment:
Same here, it's better to let things fail than to loop. At best, we should have a retry policy.

Reply:
same code twice, .. hmm I think I move the code into a function
static bool bmp180_get_convension_status(delay_ms)  { do the register test, if timed out return false }
and address your concerns.